MREUnityRuntime/MREUnityRuntimeLib/Messaging/Commands/CommandManager.cs

```diff
@@ -10,36 +10,169 @@ namespace MixedRealityExtension.Messaging.Commands
+            {
+                var commandPayloadType = commandPayload.GetType();
+                if (methodHandlers.ContainsKey(commandPayloadType))
+                {
+                    ExecuteCommandPayload(handlerContext, commandPayload, methodHandlers[commandPayloadType], onCompleteCallback);
+                }
+                else
+                {
+                    throw new Exception($"No command handler for command payload: {commandPayloadType.Name} on type: {handlerContextType.Name}");
+                }
             }
             else
             {
-                MREAPI.Logger.LogError($"No command handler registered for command payload: {commandPayloadType.Name}");
+                throw new Exception($"No command handlers registered for type: {handlerContextType.Name}");
+            }
+        }
+
+        private void ExecuteCommandPayload(ICommandHandlerContext handlerContext, ICommandPayload commandPayload, MethodInfo methodHandler, Action onCompleteCallback)
+        {
+            var handlerContextType = handlerContext.GetType();
+            var commandPayloadType = commandPayload.GetType();
+            var commandType = typeof(Command<>);
+            Type[] typeArgs = { commandPayloadType };
+            commandType = commandType.MakeGenericType(typeArgs);
+            var command = (ICommand)Activator.CreateInstance(
+                commandType,
+                new object[] {
+                            commandPayload,
+                            CreateQueuedCompletionCallback(
+                                $"{handlerContextType.Name}.{methodHandler.Name}",
+                                onCompleteCallback
+                            ) });
+            command.Execute(handlerContext, methodHandler);
+        }
+
+        public void Update()
+        {
+            // Periodically check for timed out completion callback invocations.
+            if (_pendingCompletionCallbacks.Count > 0)
+            {
+                var currTime = DateTime.Now;
+                if (_nextQueueCheckTime <= currTime)
+                {
+                    while (_pendingCompletionCallbacks.Count > 0)
+                    {
+                        // Callbacks are naturally sorted by creation time.
+                        var callback = _pendingCompletionCallbacks.First();
+                        if (callback.CreationTime <= currTime + QueuedCompletionCallbackTimeout)
+                        {
+                            // Command handler did not invoke its onCompleteCallback in a timely manner.
+                            MREAPI.Logger.LogError($"ERROR: Timeout waiting for onCompleteCallback invocation from {callback.Location}");
+                            _pendingCompletionCallbacks.Remove(callback);
+                            // PONDER: Should we forceably invoke the callback from here?
```

Comment:
If this command was executed from ActorCommandQueue.Update then the queue would stall and never process another message.

---
In reply to: [263082536](https://github.com/Microsoft/mixed-reality-extension-unity/pull/45#discussion_r263082536) [](ancestors = 263082536)

Reply:
How should we handle this case?  The PONDER here signals we should think about whether to do this, but it sounds like something is fatal if we don't based on your last comment.  Am I missing something?

Reply:
If we get here then we have a logic bug in a handler. I think the error msg is sufficient, and it names the specific handler so that's helpful.

---
In reply to: [263142163](https://github.com/Microsoft/mixed-reality-extension-unity/pull/45#discussion_r263142163) [](ancestors = 263142163)

Reply:
Sounds good.